## Part1

**You will only consider houses built from 2000 onwardsFind the most expensive houses for each year starting 2000Find the houses with the most bedrooms, bathrooms, and floors (each one separatley), for each year starting 2000.Find the houses with an area larger than the mean area of all houses (do the same for both sqft_living and sqft_lot)With respect to the above questions, explain any similarity, and/or pattern across the houses over the years.**

In [122]:
import pandas as pd
import math

In [123]:
df = pd.read_csv('kc_house_data.csv')

In [81]:
df.head(10)

id             date      price  bedrooms  bathrooms  sqft_living  \
5   7237550310  20140512T000000  1230000.0         4       4.50         5420   
9   3793500160  20150312T000000   323000.0         3       2.50         1890   
29  1873100390  20150302T000000   719000.0         4       2.50         2570   
30  8562750320  20141110T000000   580500.0         3       2.50         2320   
31  2426039314  20141201T000000   280000.0         2       1.50         1190   
40  5547700270  20140715T000000   625000.0         4       2.50         2570   
42  7203220400  20140707T000000   861990.0         5       2.75         3595   
45  8035350320  20140718T000000   488000.0         3       2.50         3160   
56  9478500640  20140819T000000   292500.0         4       2.50         2250   
57  2799800710  20150407T000000   301000.0         3       2.50         2420   

    sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
5     101930     1.0           0     0  ...     11        3890           1530   
9       6560     2.0           0     0  ...      7        1890              0   
29      7173     2.0           0     0  ...      8        2570              0   
30      3980     2.0           0     0  ...      8        2320              0   
31      1265     3.0           0     0  ...      7        1190              0   
40      5520     2.0           0     0  ...      9        2570              0   
42      5639     2.0           0     0  ...      9        3595              0   
45     13603     2.0           0     0  ...      8        3160              0   
56      4495     2.0           0     0  ...      7        2250              0   
57      4750     2.0           0     0  ...      8        2420              0   

    yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
5       2001             0    98053  47.6561 -122.005           4760   
9       2003             0    98038  47.3684 -122.031           2390   
29      2005             0    98052  47.7073 -122.110           2630   
30      2003             0    98027  47.5391 -122.070           2580   
31      2005             0    98133  47.7274 -122.357           1390   
40      2000             0    98074  47.6145 -122.027           2470   
42      2014             0    98053  47.6848 -122.016           3625   
45      2003             0    98019  47.7443 -121.977           3050   
56      2008             0    98042  47.3663 -122.114           2250   
57      2003             0    98042  47.3663 -122.122           2690   

    sqft_lot15  
5       101930  
9         7570  
29        6026  
30        3980  
31        1756  
40        5669  
42        5639  
45        9232  
56        4500  
57        4750  

[10 rows x 21 columns]

## You will only consider houses built from 2000 onwards

In [67]:
#Select houses built after 2000
df = df[df['yr_built']>=2000]
df.head()

id             date      price  bedrooms  bathrooms  sqft_living  \
5   7237550310  20140512T000000  1230000.0         4        4.5         5420   
9   3793500160  20150312T000000   323000.0         3        2.5         1890   
29  1873100390  20150302T000000   719000.0         4        2.5         2570   
30  8562750320  20141110T000000   580500.0         3        2.5         2320   
31  2426039314  20141201T000000   280000.0         2        1.5         1190   

    sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
5     101930     1.0           0     0  ...     11        3890           1530   
9       6560     2.0           0     0  ...      7        1890              0   
29      7173     2.0           0     0  ...      8        2570              0   
30      3980     2.0           0     0  ...      8        2320              0   
31      1265     3.0           0     0  ...      7        1190              0   

    yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
5       2001             0    98053  47.6561 -122.005           4760   
9       2003             0    98038  47.3684 -122.031           2390   
29      2005             0    98052  47.7073 -122.110           2630   
30      2003             0    98027  47.5391 -122.070           2580   
31      2005             0    98133  47.7274 -122.357           1390   

    sqft_lot15  
5       101930  
9         7570  
29        6026  
30        3980  
31        1756  

[5 rows x 21 columns]

In [68]:
df.shape[0]

4759

**We have 4759 house built after 2000**

In [69]:
years = df['yr_built'].unique()
print(sorted(years))

[2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015]


**The building year is from 2000 to 2015.**

## 1. Find the most expensive houses for each year starting 2000

In [70]:
for year in range(2000, 2016):
    most_expensive_house = df.loc[(df["yr_built"] == year) & (df["price"] == df[df["yr_built"] == year]["price"].max())]
    
    if not most_expensive_house.empty:
        print(f"Year: {year}")
        print(most_expensive_house)
        print("\n")

Year: 2000
               id             date      price  bedrooms  bathrooms  \
16302  7397300170  20140530T000000  3710000.0         4        3.5   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
16302         5550     28078     2.0           0     2  ...     12   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
16302        3350           2200      2000             0    98039  47.6395   

          long  sqft_living15  sqft_lot15  
16302 -122.234           2980       19602  

[1 rows x 21 columns]


Year: 2001
              id             date      price  bedrooms  bathrooms  \
9254  9208900037  20140919T000000  6890000.0         6       7.75   

      sqft_living  sqft_lot  floors  waterfront  view  ...  grade  sqft_above  \
9254         9890     31374     2.0           0     4  ...     13        8860   

      sqft_basement  yr_built  yr_renovated  zipcode      lat    long  \
9254           1030      2001             0    9803

**Year: 2000,Price:3710000 <br>
Year: 2001, Price: 6890000<br>
Year: 2002, Price: 3080000<br>
Year: 2003, Price: 4210000<br>
Year: 2004, Price: 4500000<br>
Year: 2005, Price: 3640000<br>
Year: 2006, Price: 3100000<br>
Year: 2007, Price: 3650000<br>
Year: 2000, Price: 4000000<br>
Year: 2000, Price: 5350000<br>
Year: 2000, Price: 2950000<br>
Year: 2000, Price: 2580000<br>
Year: 2000, Price: 2000000<br>
Year: 2000, Price: 3000000<br>
Year: 2000, Price: 3420000<br>
Year: 2000, Price: 2000000<br>**

## 2. Find the houses with the most bedrooms, bathrooms, and floors (each one separatley), for each year starting 2000.

In [71]:
for year in range(2000, 2016):
    most_bedrooms_house = df.loc[(df["yr_built"] == year) & (df["bedrooms"] == df[df["yr_built"] == year]["bedrooms"].max())]
    most_bathrooms_house = df.loc[(df["yr_built"] == year) & (df["bathrooms"] == df[df["yr_built"] == year]["bathrooms"].max())]
    most_floors_house = df.loc[(df["yr_built"] == year) & (df["floors"] == df[df["yr_built"] == year]["floors"].max())]
    
    if not most_bedrooms_house.empty:
        print(f"Year: {year} - Most Bedrooms:")
        print(most_bedrooms_house)
        print("\n")
        
    if not most_bathrooms_house.empty:
        print(f"Year: {year} - Most Bathrooms:")
        print(most_bathrooms_house)
        print("\n")
        
    if not most_floors_house.empty:
        print(f"Year: {year} - Most Floors:")
        print(most_floors_house)
        print("\n")

Year: 2000 - Most Bedrooms:
              id             date      price  bedrooms  bathrooms  \
2668  4139500200  20150305T000000  1380000.0         6        4.5   
7681  9286000150  20150330T000000  1130000.0         6        4.0   

      sqft_living  sqft_lot  floors  waterfront  view  ...  grade  sqft_above  \
2668         5740     10312     2.0           0     2  ...     11        3610   
7681         5330     18116     2.0           0     0  ...     11        3950   

      sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
2668           2130      2000             0    98006  47.5533 -122.110   
7681           1380      2000             0    98006  47.5503 -122.137   

      sqft_living15  sqft_lot15  
2668           4350       11917  
7681           4590       16900  

[2 rows x 21 columns]


Year: 2000 - Most Bathrooms:
              id             date      price  bedrooms  bathrooms  \
6501  2626069030  20150209T000000  1940000.0         4       5.75   

  

Year: 2007 - Most Bedrooms:
               id             date      price  bedrooms  bathrooms  \
2345   3185600055  20140611T000000   495000.0         6       5.00   
7415   5017000470  20140710T000000  1980000.0         6       4.50   
20578   424069279  20150328T000000  1180000.0         6       6.50   
20838  3425069117  20140828T000000  1280000.0         6       5.25   
21186  7896300592  20150114T000000   303500.0         6       4.50   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
2345          3440      4500     2.0           0     0  ...      8   
7415          4800      9097     2.0           0     0  ...     10   
20578         6260     10955     2.0           0     0  ...     11   
20838         6160     27490     2.0           0     0  ...     11   
21186         3390      7200     2.0           0     0  ...      8   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
2345         3280            160      2007          

Year: 2012 - Most Bedrooms:
               id             date     price  bedrooms  bathrooms  \
19793  5609000311  20150410T000000  729999.0         6       4.50   
20873  8946390040  20140508T000000  375000.0         6       2.25   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
19793         3600      6110     2.0           0     0  ...      9   
20873         3206      5793     2.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
19793        2510           1090      2012             0    98118  47.5687   
20873        3206              0      2012             0    98032  47.3690   

          long  sqft_living15  sqft_lot15  
19793 -122.291           1360        5800  
20873 -122.287           2527        5804  

[2 rows x 21 columns]


Year: 2012 - Most Bathrooms:
               id             date     price  bedrooms  bathrooms  \
6389     42000130  20140924T000000  600000.0         5        4

**2000, bedrooms: 6(grade:11), bathrooms: 5.75 (grade:12), floors: 3 <br>
2001, bedrooms: 7(grade:10), bathrooms: 6(grade:13), floors: 3 <br>
2002, bedrooms: 6(grade:10), bathrooms: 5(grade:10), floors: 3 <br>
2003, bedrooms: 6(grade:from 8 to 10), bathrooms: 5(grade:12), floors: 3 <br>
2004, bedrooms: 7(grade:10), bathrooms: 5.75(grade:12), floors: 3 <br>
2005, bedrooms: 6(grade:10), bathrooms: 6(grade:12), floors: 3.5 <br>
2006, bedrooms: 6(grade:7), bathrooms: 5.5(grade:11), floors: 3.5 <br>
2007, bedrooms: 6(grade:from 8 to 11), bathrooms: 6.5(grade:11), floors: 3.5 <br>
2008, bedrooms: 10, bathrooms: 5.5, floors: 3.5 <br>
2009, bedrooms: 6, bathrooms: 5, floors: 3 <br>
2010, bedrooms: 5, bathrooms: 5, floors: 3 <br>
2011, bedrooms: 5, bathrooms: 4, floors: 3 <br>
2012, bedrooms: 6, bathrooms: 4.5, floors: 3 <br>
2013, bedrooms: 7, bathrooms: 5, floors: 3 <br>
2014, bedrooms: 6, bathrooms: 5, floors: 3 <br>
2015, bedrooms: 5, bathrooms: 4.5, floors: 3 <br>**

## 3. Find the houses with an area larger than the mean area of all houses (do the same for both sqft_living and sqft_lot)

In [72]:
# Calculate the mean area
mean_living = df['sqft_living'].mean()
mean_lot = df['sqft_lot'].mean()

In [73]:
# Select houses with linving area larger than the mean area
large_sqft_living = df.loc[df['sqft_living']>mean_living]
print(large_sqft_living)

               id             date      price  bedrooms  bathrooms  \
5      7237550310  20140512T000000  1230000.0         4       4.50   
29     1873100390  20150302T000000   719000.0         4       2.50   
40     5547700270  20140715T000000   625000.0         4       2.50   
42     7203220400  20140707T000000   861990.0         5       2.75   
45     8035350320  20140718T000000   488000.0         3       2.50   
...           ...              ...        ...       ...        ...   
21598  8956200760  20141013T000000   541800.0         4       2.50   
21599  7202300110  20140915T000000   810000.0         4       3.00   
21600   249000205  20141015T000000  1540000.0         5       3.75   
21605  3448900210  20141014T000000   610685.0         4       2.50   
21606  7936000429  20150326T000000  1010000.0         4       3.50   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
5             5420    101930     1.0           0     0  ...     11   
29            2570 

In [74]:
# Select houses with lot area larger than the mean area
large_sqft_lot = df.loc[df['sqft_lot']>mean_lot]
print(large_sqft_lot)

               id             date      price  bedrooms  bathrooms  \
5      7237550310  20140512T000000  1230000.0         4        4.5   
45     8035350320  20140718T000000   488000.0         3        2.5   
79     1531000030  20150323T000000   720000.0         4        2.5   
114    8961960160  20141028T000000   480000.0         4        2.5   
147    3224079105  20140806T000000   430000.0         2        2.5   
...           ...              ...        ...       ...        ...   
21532   324069112  20140617T000000  1330000.0         4        4.0   
21540   715010530  20150113T000000  1880000.0         5        3.5   
21548  8835770330  20140819T000000  1060000.0         2        1.5   
21551  1561750040  20141224T000000  1380000.0         5        4.5   
21553  2522059251  20150409T000000   465000.0         3        2.5   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
5             5420    101930     1.0           0     0  ...     11   
45            3160 

## 4. Similarity, and/or pattern across the houses over the years

1. **The most expensive houses got more expensive from 2000 to 2007, but then prices went up and down. This could be because of the economy, the housing market.**
2.  **The houses with the most bedrooms, bathrooms, and floors tend to have relatively high grades. <br>
      There was a trend of building houses with more bedrooms, bathrooms, and floors in the earlier years (2000-2008).**
3. **Houses with larger areas than the mean area of all houses for both sqft_living and sqft_lot appear to be spread across different years. This indicates that larger properties are not limited to any specific time perio.**

# Part 2

## 1. Create a function which takes in a set of coordinates (lat and long) and another set, and calculates the distance between them.

In [112]:
df.head()

id             date      price  bedrooms  bathrooms  sqft_living  \
5   7237550310  20140512T000000  1230000.0         4        4.5         5420   
9   3793500160  20150312T000000   323000.0         3        2.5         1890   
29  1873100390  20150302T000000   719000.0         4        2.5         2570   
30  8562750320  20141110T000000   580500.0         3        2.5         2320   
31  2426039314  20141201T000000   280000.0         2        1.5         1190   

    sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
5     101930     1.0           0     0  ...     11        3890           1530   
9       6560     2.0           0     0  ...      7        1890              0   
29      7173     2.0           0     0  ...      8        2570              0   
30      3980     2.0           0     0  ...      8        2320              0   
31      1265     3.0           0     0  ...      7        1190              0   

    yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
5       2001             0    98053  47.6561 -122.005           4760   
9       2003             0    98038  47.3684 -122.031           2390   
29      2005             0    98052  47.7073 -122.110           2630   
30      2003             0    98027  47.5391 -122.070           2580   
31      2005             0    98133  47.7274 -122.357           1390   

    sqft_lot15  
5       101930  
9         7570  
29        6026  
30        3980  
31        1756  

[5 rows x 21 columns]

In [118]:
def distance_between_coordinates(lat_1, lon_1,lat_2, lon_2):
    lat_1= lat_1.astype(float) #convert to float
    long_1= long_1.astype(float)
    lat_2= lat_2.astype(float)
    long_2= long_2.astype(float)

    radius_of_earth = 6371.01

    d_lat = lat_2 - lat_1
    d_lon = lon_2 - lon_1

    a = (d_lat**2) + (d_lon**2)
    c = math.sqrt(a)

    distance = radius_of_earth * c

    return distance


# 2. You need to include a check to make sure that any given ID actually exists.

In [119]:
def distance_between_houses(id1,id2):
    # Select houses
    house_1 = df.loc[df['id']==id1]
    house_2 = df.loc[df['id']==id2]
    
    #Check if ids are valide 
    #if house_1.empty or house_2.empty:
        #return "Invalid house ID(s)"
    
    # Get coord
    lat_1 = house_1['lat']
    long_1 = house_2['long']
    lat_2 = house_2['lat'] 
    long_2 = house_2['long']
    
    # Calculate distance
    d = distance_between_coordinates(lat_1,long_1, lat_2, long_2)
    
    #Return distance
    return d

In [120]:
id_1 = 3793500160
id_2 = 2426039314
d = distance_between_houses(id_1,id_2)
print(d)

UnboundLocalError: cannot access local variable 'long_1' where it is not associated with a value

# 3. Given this function, try to find similarities between distance from one another and price.

## Part 3

## Research the usage of .at and .iat, include examples and explanations using the housing dataset and custom datasets you created

The .at and .iat methods in pandas are used to access individual elements within a DataFrame or Series based on their labels or integer positions, respectively. These methods are designed for fast and efficient single-item access. Here's how they work along with examples using the housing dataset and a custom dataset:

**.at Method: <br>
The .at method is used to access a single element in a DataFrame using row and column labels.** <br>
- df.at[row_label, column_label]


In [124]:
price0 = df.at[0, 'price']
print("Price of house at index 0:", price0)

Price of house at index 0: 221900.0


**.iat Method:<br>
The .iat method is used to access a single element in a DataFrame using integer positions for both rows and columns.**<br>
- df.iat[row_position, column_position]

In [126]:
element01 = df.iat[0, 1]
print("Element at (0, 1):", element01)

Element at (0, 1): 20141013T000000
